File: cpp/strings/suffix_automaton.cpp
#include <bits/stdc++.h>
using namespace std;
struct SAM
{
  struct S
  {
    int link = -1, len = 0;
    map<char, int> nxt;
    long long occ = 0;
  };
  vector<S> st;
  int last = 0;
  SAM(int n = 0)
  {
    st.reserve(2 * n);
    st.push_back(S());
  }
  void extend(char c)
  {
    int cur = st.size();
    st.push_back(S());
    st[cur].len = st[last].len + 1;
    st[cur].occ = 1;
    int p = last;
    while (p != -1 && !st[p].nxt.count(c))
      st[p].nxt[c] = cur, p = st[p].link;
    if (p == -1)
      st[cur].link = 0;
    else
    {
      int q = st[p].nxt[c];
      if (st[p].len + 1 == st[q].len)
        st[cur].link = q;
      else
      {
        int cl = st.size();
        st.push_back(st[q]);
        st[cl].len = st[p].len + 1;
        st[cl].occ = 0;
        while (p != -1 && st[p].nxt[c] == q)
          st[p].nxt[c] = cl, p = st[p].link;
        st[q].link = st[cur].link = cl;
      }
    }
    last = cur;
  }
  void build(const string &s)
  {
    for (char c : s)
      extend(c);
  }
  long long distinct()
  {
    long long ans = 0;
    for (int i = 1; i < (int)st.size(); i++)
      ans += st[i].len - st[st[i].link].len;
    return ans;
  }
};

File: python/strings/suffix_automaton.py
class SuffixAutomaton:
    def __init__(self):
        self.next = [{}]
        self.link = [-1]
        self.length = [0]
        self.last = 0

    def extend(self, ch):
        cur = len(self.next)
        self.next.append({})
        self.length.append(self.length[self.last] + 1)
        self.link.append(0)
        p = self.last
        while p != -1 and ch not in self.next[p]:
            self.next[p][ch] = cur
            p = self.link[p]
        if p == -1:
            self.link[cur] = 0
        else:
            q = self.next[p][ch]
            if self.length[p] + 1 == self.length[q]:
                self.link[cur] = q
            else:
                clone = len(self.next)
                self.next.append(self.next[q].copy())
                self.length.append(self.length[p] + 1)
                self.link.append(self.link[q])
                while p != -1 and self.next[p].get(ch) == q:
                    self.next[p][ch] = clone
                    p = self.link[p]
                self.link[q] = self.link[cur] = clone
        self.last = cur

    @classmethod
    def from_string(cls, s):
        sam = cls()
        for ch in s:
            sam.extend(ch)
        return sam
